### ISSUE TITLE
CRS related error message when plotting zones with recent versions of sf

### ISSUE THREAD

**Issue by Robinlovelace** [mtennekes/tmap#492]
I think this is the same issue as https://github.com/Robinlovelace/geocompr/issues/566. I've just bumped into it, enabling a simple reproducible example to illustrate the problem. Interesting that it works intermittently: the first, smaller, `z` object plots fine but the second one only works with `sf` and `mapview()` plotting methods:

```r
remotes::install_github("mtennekes/tmap")
library(tmap)
u = "https://github.com/npct/pct-outputs-regional-notR/raw/master/commute/lsoa/isle-of-wight/z.geojson"
z = sf::st_read(u)
plot(z["bicycle"])
qtm(z)
tm_shape(z) +
  tm_fill("bicycle", palette = "RdYlBu") 
tmap_mode("view")
qtm(z)

# another region
tmap_mode("plot")
u = "https://github.com/npct/pct-outputs-regional-notR/raw/master/commute/lsoa/west-yorkshire/z.geojson"
z = sf::st_read(u)
plot(z["bicycle"])
qtm(z)
tm_shape(z) +
  tm_fill("bicycle", palette = "RdYlBu") 
tmap_mode("view")
qtm(z)
mapview::mapview(z)

devtools::session_info()
```

**Comment by Robinlovelace** [comment]
Output from the previous commands on my laptop:

``` r
remotes::install_github("mtennekes/tmap")
#> Using github PAT from envvar GITHUB_PAT
#> Skipping install of 'tmap' from a github remote, the SHA1 (73ddc01f) has not changed since last install.
#>   Use `force = TRUE` to force installation
library(tmap)
u = "https://github.com/npct/pct-outputs-regional-notR/raw/master/commute/lsoa/isle-of-wight/z.geojson"
z = sf::st_read(u)
#> Reading layer `z' from data source `https://github.com/npct/pct-outputs-regional-notR/raw/master/commute/lsoa/isle-of-wight/z.geojson' using driver `GeoJSON'
#> Simple feature collection with 89 features and 129 fields
#> geometry type:  MULTIPOLYGON
#> dimension:      XY
#> bbox:           xmin: -1.591613 ymin: 50.57491 xmax: -1.070027 ymax: 50.76684
#> geographic CRS: WGS 84
plot(z["bicycle"])
```

![](https://i.imgur.com/5ROA5J8.png)

``` r
qtm(z)
```

![](https://i.imgur.com/RDI52R1.png)

``` r
tm_shape(z) +
  tm_fill("bicycle", palette = "RdYlBu") 
```

![](https://i.imgur.com/IgBBDAO.png)

``` r
tmap_mode("view")
#> tmap mode set to interactive viewing
qtm(z)
```

![](https://i.imgur.com/yy8Pgow.png)

``` r

# another region
tmap_mode("plot")
#> tmap mode set to plotting
u = "https://github.com/npct/pct-outputs-regional-notR/raw/master/commute/lsoa/west-yorkshire/z.geojson"
z = sf::st_read(u)
#> Reading layer `z' from data source `https://github.com/npct/pct-outputs-regional-notR/raw/master/commute/lsoa/west-yorkshire/z.geojson' using driver `GeoJSON'
#> Simple feature collection with 1388 features and 129 fields
#> geometry type:  MULTIPOLYGON
#> dimension:      XY
#> bbox:           xmin: -2.173294 ymin: 53.51991 xmax: -1.198818 ymax: 53.96139
#> geographic CRS: WGS 84
plot(z["bicycle"])
```

![](https://i.imgur.com/cM12XwT.png)

``` r
qtm(z)
#> Linking to GEOS 3.8.0, GDAL 3.0.4, PROJ 6.3.1
#> Error in CPL_lwgeom_transform(x, crs): st_lwgeom_transform: one of the proj strings is invalid
tm_shape(z) +
  tm_fill("bicycle", palette = "RdYlBu") 
#> Error in CPL_lwgeom_transform(x, crs): st_lwgeom_transform: one of the proj strings is invalid
tmap_mode("view")
#> tmap mode set to interactive viewing
qtm(z)
#> Error in CPL_lwgeom_transform(x, crs): st_lwgeom_transform: one of the proj strings is invalid
mapview::mapview(z)
```

![](https://i.imgur.com/AxNNvho.png)

``` r

devtools::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 4.0.2 (2020-06-22)
#>  os       Ubuntu 20.04.1 LTS          
#>  system   x86_64, linux-gnu           
#>  ui       X11                         
#>  language en_GB:en                    
#>  collate  en_GB.UTF-8                 
#>  ctype    en_GB.UTF-8                 
#>  tz       Europe/London               
#>  date     2020-08-25                  
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package           * version    date       lib
#>  abind               1.4-5      2016-07-21 [3]
#>  assertthat          0.2.1      2019-03-21 [3]
#>  backports           1.1.9      2020-08-24 [3]
#>  base64enc           0.1-3      2015-07-28 [3]
#>  brew                1.0-6      2011-04-13 [3]
#>  callr               3.4.3      2020-03-28 [3]
#>  class               7.3-17     2020-04-26 [4]
#>  classInt            0.4-3      2020-04-07 [3]
#>  cli                 2.0.2      2020-02-28 [3]
#>  codetools           0.2-16     2018-12-24 [4]
#>  colorspace          1.4-1      2019-03-18 [3]
#>  crayon              1.3.4      2017-09-16 [3]
#>  crosstalk           1.1.0.1    2020-03-13 [3]
#>  curl                4.3        2019-12-02 [3]
#>  DBI                 1.1.0      2019-12-15 [3]
#>  desc                1.2.0      2018-05-01 [3]
#>  devtools            2.3.1      2020-07-21 [3]
#>  dichromat           2.0-0      2013-01-24 [1]
#>  digest              0.6.25     2020-02-23 [3]
#>  dplyr               1.0.2      2020-08-18 [3]
#>  e1071               1.7-3      2019-11-26 [3]
#>  ellipsis            0.3.1      2020-05-15 [3]
#>  evaluate            0.14       2019-05-28 [3]
#>  fansi               0.4.1      2020-01-08 [3]
#>  fs                  1.5.0      2020-07-31 [3]
#>  gdtools             0.2.2      2020-04-03 [3]
#>  generics            0.0.2      2018-11-29 [3]
#>  glue                1.4.1      2020-05-13 [3]
#>  highr               0.8        2019-03-20 [3]
#>  htmltools           0.5.0      2020-06-16 [3]
#>  htmlwidgets         1.5.1      2019-10-08 [3]
#>  jsonlite            1.7.0      2020-06-25 [3]
#>  KernSmooth          2.23-17    2020-04-26 [4]
#>  knitr               1.29       2020-06-23 [3]
#>  lattice             0.20-41    2020-04-02 [4]
#>  leafem              0.1.3      2020-07-26 [3]
#>  leaflet             2.0.3      2019-11-16 [3]
#>  leaflet.providers   1.9.0      2019-11-09 [3]
#>  leafpop             0.0.5      2019-12-09 [3]
#>  leafsync            0.1.0      2019-03-05 [1]
#>  lifecycle           0.2.0      2020-03-06 [3]
#>  lwgeom              0.2-5      2020-06-12 [1]
#>  magrittr            1.5        2014-11-22 [3]
#>  mapview             2.9.0      2020-08-11 [3]
#>  memoise             1.1.0      2017-04-21 [3]
#>  munsell             0.5.0      2018-06-12 [3]
#>  pillar              1.4.6      2020-07-10 [3]
#>  pkgbuild            1.1.0      2020-07-13 [3]
#>  pkgconfig           2.0.3      2019-09-22 [3]
#>  pkgload             1.1.0      2020-05-29 [3]
#>  png                 0.1-7      2013-12-03 [3]
#>  prettyunits         1.1.1      2020-01-24 [3]
#>  processx            3.4.3      2020-07-05 [3]
#>  ps                  1.3.4      2020-08-11 [3]
#>  purrr               0.3.4      2020-04-17 [3]
#>  R6                  2.4.1      2019-11-12 [3]
#>  raster              3.3-13     2020-07-17 [3]
#>  RColorBrewer        1.1-2      2014-12-07 [3]
#>  Rcpp                1.0.5      2020-07-06 [1]
#>  remotes             2.2.0      2020-07-21 [3]
#>  rlang               0.4.7      2020-07-09 [3]
#>  rmarkdown           2.3        2020-06-18 [3]
#>  rprojroot           1.3-2      2018-01-03 [3]
#>  s2                  1.0.2      2020-08-03 [1]
#>  satellite           1.0.2      2019-12-09 [3]
#>  scales              1.1.1      2020-05-11 [3]
#>  sessioninfo         1.1.1      2018-11-05 [3]
#>  sf                * 0.9-6      2020-08-25 [1]
#>  sp                  1.4-4      2020-08-25 [1]
#>  stars               0.4-3      2020-07-08 [1]
#>  stringi             1.4.6      2020-02-17 [3]
#>  stringr             1.4.0      2019-02-10 [3]
#>  svglite             1.2.3.2    2020-07-07 [3]
#>  systemfonts         0.2.3      2020-06-09 [3]
#>  testthat            2.3.2      2020-03-02 [3]
#>  tibble              3.0.3      2020-07-10 [3]
#>  tidyselect          1.1.0      2020-05-11 [3]
#>  tmap              * 3.2        2020-08-25 [1]
#>  tmaptools           3.1        2020-07-03 [1]
#>  units               0.6-7      2020-06-13 [3]
#>  usethis             1.6.1.9001 2020-08-19 [1]
#>  uuid                0.1-4      2020-02-26 [3]
#>  vctrs               0.3.2      2020-07-15 [3]
#>  viridisLite         0.3.0      2018-02-01 [3]
#>  webshot             0.5.2      2019-11-22 [3]
#>  withr               2.2.0      2020-04-20 [3]
#>  wk                  0.3.2      2020-08-03 [1]
#>  xfun                0.16       2020-07-24 [3]
#>  XML                 3.99-0.5   2020-07-23 [1]
#>  yaml                2.2.1      2020-02-01 [3]
#>  source                              
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.2)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.2)                      
#>  CRAN (R 4.0.2)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.2)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.2)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.1)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.2)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.2)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.2)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.2)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.2)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.2)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.2)                      
#>  CRAN (R 4.0.2)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.2)                      
#>  CRAN (R 4.0.2)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.2)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.2)                      
#>  CRAN (R 4.0.2)                      
#>  CRAN (R 4.0.2)                      
#>  CRAN (R 4.0.1)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.2)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  Github (r-spatial/sf@f36719c)       
#>  Github (rsbivand/sp@3a47130)        
#>  CRAN (R 4.0.2)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.2)                      
#>  CRAN (R 4.0.1)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.2)                      
#>  CRAN (R 4.0.0)                      
#>  Github (mtennekes/tmap@73ddc01)     
#>  Github (mtennekes/tmaptools@947f3bd)
#>  CRAN (R 4.0.1)                      
#>  Github (r-lib/usethis@860c1ea)      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.2)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.0)                      
#>  CRAN (R 4.0.2)                      
#>  CRAN (R 4.0.2)                      
#>  CRAN (R 4.0.2)                      
#>  CRAN (R 4.0.0)                      
#> 
#> [1] /home/robin/R/x86_64-pc-linux-gnu-library/4.0
#> [2] /usr/local/lib/R/site-library
#> [3] /usr/lib/R/site-library
#> [4] /usr/lib/R/library
```

<sup>Created on 2020-08-25 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by Robinlovelace** [comment]
Good news: it seems the bug only affects set-ups with certain versions of PROJ installed. With PROJ 6.3.1 it fails, as shown above. With the more recent version of PROJ it works :tada:  as shown below. I am not sure how many computers have versions of PROJ that cause this issue but I think this is getting close to the route cause of the issue and suggests that we should be using the latest version of PROJ in the geocompr repo and elsewhere. Hope this is of use/interest, it certainly confused me!

**Comment by Robinlovelace** [comment]
``` r
remotes::install_github("mtennekes/tmap")
#> Using github PAT from envvar GITHUB_PAT
#> Skipping install of 'tmap' from a github remote, the SHA1 (73ddc01f) has not changed since last install.
#>   Use `force = TRUE` to force installation
    library(tmap)
    u = "https://github.com/npct/pct-outputs-regional-notR/raw/master/commute/lsoa/isle-of-wight/z.geojson"
    z = sf::st_read(u)
#> Reading layer `z' from data source `https://github.com/npct/pct-outputs-regional-notR/raw/master/commute/lsoa/isle-of-wight/z.geojson' using driver `GeoJSON'
#> Simple feature collection with 89 features and 129 fields
#> geometry type:  MULTIPOLYGON
#> dimension:      XY
#> bbox:           xmin: -1.591613 ymin: 50.57491 xmax: -1.070027 ymax: 50.76684
#> geographic CRS: WGS 84
    plot(z["bicycle"])
```

![](https://i.imgur.com/7CjZsOX.png)

``` r
    qtm(z)
#> Linking to GEOS 3.8.0, GDAL 3.0.4, PROJ 7.0.0
```

<details>

![](https://i.imgur.com/C5IkCYE.png)

``` r
    tm_shape(z) +
        tm_fill("bicycle", palette = "RdYlBu") 
```

![](https://i.imgur.com/IxZjAKd.png)

``` r
    tmap_mode("view")
#> tmap mode set to interactive viewing
    qtm(z)
```

![](https://i.imgur.com/zc5hoXI.png)

``` r
    
    # another region
    tmap_mode("plot")
#> tmap mode set to plotting
    u = "https://github.com/npct/pct-outputs-regional-notR/raw/master/commute/lsoa/west-yorkshire/z.geojson"
    z = sf::st_read(u)
#> Reading layer `z' from data source `https://github.com/npct/pct-outputs-regional-notR/raw/master/commute/lsoa/west-yorkshire/z.geojson' using driver `GeoJSON'
#> Simple feature collection with 1388 features and 129 fields
#> geometry type:  MULTIPOLYGON
#> dimension:      XY
#> bbox:           xmin: -2.173294 ymin: 53.51991 xmax: -1.198818 ymax: 53.96139
#> geographic CRS: WGS 84
    plot(z["bicycle"])
```

![](https://i.imgur.com/yDTsZcq.png)

``` r
    qtm(z)
```

![](https://i.imgur.com/CmmX7mP.png)

``` r
    tm_shape(z) +
        tm_fill("bicycle", palette = "RdYlBu") 
```

![](https://i.imgur.com/n9JO1mX.png)

``` r
    tmap_mode("view")
#> tmap mode set to interactive viewing
    qtm(z)
```

![](https://i.imgur.com/IeaIzTS.png)

``` r
    mapview::mapview(z)
#> Error in loadNamespace(name): there is no package called 'mapview'
```

<sup>Created on 2020-08-25 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>



<summary>Session info</summary>

``` r
sessionInfo()
#> R version 4.0.2 (2020-06-22)
#> Platform: x86_64-pc-linux-gnu (64-bit)
#> Running under: Ubuntu 18.04.5 LTS
#> 
#> Matrix products: default
#> BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1
#> LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.7.1
#> 
#> locale:
#>  [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C              
#>  [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8    
#>  [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   
#>  [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                 
#>  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
#> [11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       
#> 
#> attached base packages:
#> [1] stats     graphics  grDevices utils     datasets  methods   base     
#> 
#> other attached packages:
#> [1] sf_0.9-5 tmap_3.2
#> 
#> loaded via a namespace (and not attached):
#>  [1] Rcpp_1.0.5              lattice_0.20-41         png_0.1-7              
#>  [4] class_7.3-17            ps_1.3.4                leaflet.providers_1.9.0
#>  [7] digest_0.6.25           mime_0.9                R6_2.4.1               
#> [10] evaluate_0.14           e1071_1.7-3             httr_1.4.2             
#> [13] highr_0.8               pillar_1.4.6            rlang_0.4.7            
#> [16] curl_4.3                callr_3.4.3             raster_3.3-13          
#> [19] rmarkdown_2.3           webshot_0.5.2           stringr_1.4.0          
#> [22] htmlwidgets_1.5.1       compiler_4.0.2          xfun_0.16              
#> [25] pkgconfig_2.0.3         tmaptools_3.1           base64enc_0.1-3        
#> [28] htmltools_0.5.0         tidyselect_1.1.0        tibble_3.0.3           
#> [31] codetools_0.2-16        XML_3.99-0.5            viridisLite_0.3.0      
#> [34] crayon_1.3.4            dplyr_1.0.2             grid_4.0.2             
#> [37] jsonlite_1.7.0          lwgeom_0.2-5            lifecycle_0.2.0        
#> [40] DBI_1.1.0               magrittr_1.5            units_0.6-7            
#> [43] KernSmooth_2.23-17      stringi_1.4.6           leafsync_0.1.0         
#> [46] remotes_2.2.0           leaflet_2.0.3           sp_1.4-2               
#> [49] xml2_1.3.2              ellipsis_0.3.1          generics_0.0.2         
#> [52] vctrs_0.3.2             RColorBrewer_1.1-2      tools_4.0.2            
#> [55] dichromat_2.0-0         leafem_0.1.3            glue_1.4.1             
#> [58] purrr_0.3.4             crosstalk_1.1.0.1       processx_3.4.3         
#> [61] abind_1.4-5             parallel_4.0.2          yaml_2.2.1             
#> [64] stars_0.4-3             classInt_0.4-3          knitr_1.29
```

</details>

**Comment by mtennekes** [comment]
Thanks @Robinlovelace !

I got the same error with `Linking to GEOS 3.8.0, GDAL 3.0.4, PROJ 6.3.1`. Short reproducible example:

```r
library(tmap)
data(World)
qtm(World, projection = 4326)
#> proj_create: several objects matching this name: Nouvelle Triangulation Francaise Paris grades to 4326, 
#> Nouvelle Triangulation Francaise Paris grades to 4326 Error in CPL_lwgeom_transform(x, crs) : 
#>  st_lwgeom_transform: one of the proj strings is invalid
```


In `tmaptools::approx_areas`, there is the following code chunk:

```r
areas <- tryCatch(sf::st_area(shp),
                  error = function(e) {
                      lwgeom::st_geod_area(lwgeom::st_transform_proj(shp, crs = 4326))
                  })
```

In `tmap` I need to calculate polygon areas for a few reasons, including: removing small polygons, and calculating density values. 

When `shp` is unprojected (i.e. epsg 4326), `sf::st_area(shp)` fails with the message `package s2 required, please install it first`. In the code above I use lwgeom as a fallback option, but it breaks, probably due to a PROJ issue, which seems to be solved in PROJ > 6.3.1.

@edzer Great that the `s2` package has already been published! Probably, it makes sense let tmaptools import it, for `approx_areas` and `approx_distances` (which currently also uses lwgeom as fallback for `sf::st_distance`).
